src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/AssetsFileTopLevelDependenciesCollectionSourceProvider.cs

```diff
@@ -0,0 +1,144 @@
+
+using NuGet.VisualStudio.SolutionExplorer.Models;
+
+namespace NuGet.VisualStudio.SolutionExplorer
+{
+    /// <summary>
+    /// Base class for attaching children to top-level dependencies that come from the assets file.
+    /// </summary>
+    /// <remarks>
+    /// Templates out common code with a bunch of protected methods to override for specific item types.
+    /// </remarks>
+    internal abstract class AssetsFileTopLevelDependenciesCollectionSourceProvider<TIdentity, TItem> : DependenciesAttachedCollectionSourceProviderBase
+        where TItem : class, IRelatableItem
+    {
+        protected AssetsFileTopLevelDependenciesCollectionSourceProvider(ProjectTreeFlags flags)
+            : base(flags)
+        {
+        }
+
+        protected abstract bool TryGetIdentity(Properties properties, out TIdentity identity);
+
+        protected abstract bool TryGetLibrary(AssetsFileTarget target, TIdentity identity, [NotNullWhen(returnValue: true)] out AssetsFileTargetLibrary? library);
+
+        protected abstract TItem CreateItem(AssetsFileTarget targetData, AssetsFileTargetLibrary library);
+
+        protected abstract bool TryUpdateItem(TItem item, AssetsFileTarget targetData, AssetsFileTargetLibrary library);
+
+        protected override bool TryCreateCollectionSource(
+            IVsHierarchyItem hierarchyItem,
+            string flagsString,
+            string? target,
+            IRelationProvider relationProvider,
+            [NotNullWhen(returnValue: true)] out AggregateRelationCollectionSource? containsCollectionSource)
+        {
+            TIdentity identity;
+
+            if (!ErrorHandler.Succeeded(hierarchyItem.HierarchyIdentity.Hierarchy.GetProperty(
+                hierarchyItem.HierarchyIdentity.ItemID, (int)__VSHPROPID.VSHPROPID_ExtObject, out object projectItemObject)))
+            {
+                containsCollectionSource = null;
+                return false;
+            }
+            else
+            {
+                Properties? properties = (projectItemObject as ProjectItem)?.Properties;
+
+                if (properties == null || !TryGetIdentity(properties, out identity))
+                {
+                    containsCollectionSource = null;
+                    return false;
+                }
+            }
+
+            UnconfiguredProject? unconfiguredProject = hierarchyItem.HierarchyIdentity.Hierarchy.AsUnconfiguredProject();
+
+            // Find the data source
+            IAssetsFileDependenciesDataSource? dataSource = unconfiguredProject?.Services.ExportProvider.GetExportedValueOrDefault<IAssetsFileDependenciesDataSource>();
+
+            if (unconfiguredProject == null || dataSource == null)
+            {
+                containsCollectionSource = null;
+                return false;
+            }
+
+            IProjectThreadingService projectThreadingService = unconfiguredProject.Services.ThreadingPolicy;
+
+            projectThreadingService.VerifyOnUIThread();
```

Comment:
We have banned `ThreadHelper` in the .NET Project System because it's static and prohibits unit testing.

https://github.com/dotnet/project-system/blob/1587349ca69b35ce320f7af8f264e89ce12bdea6/src/Common/BannedSymbols.txt#L82